teamcity-rest-client-impl/src/main/kotlin/org/jetbrains/teamcity/rest/coroutines/implementation.kt

```diff
@@ -160,10 +157,29 @@ private fun selectRestApiCountForPagedRequests(limitResults: Int?, pageSize: Int
     return pageSize ?: limitResults?.let { min(it, reasonableMaxPageSize) }
 }
 
+private class NodeIdCookieJar(val nodeId: String) : CookieJar {
+    override fun loadForRequest(url: HttpUrl): List<Cookie> {
+        val domain = url.topPrivateDomain() ?: return emptyList()
+
+        return listOf(
+            Cookie.Builder()
+                .domain(domain)
+                .name("X-TeamCity-Node-Id-Cookie")
+                .value(nodeId)
+                .build()
+        )
+    }
+
+    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
+        // noop, same as CookieJar.NO_COOKIES
+    }
+}
```

Comment:
User don't need to always configure node-id on client side. If not set explicitly, node-id have to be taken from TeamCity response, `Set-Cookie` header

So the flow is
1. Client sends a first request without node-id
2. Client receives a response with `Set-Cookie: X-TeamCity-Node-Id-Cookie=node-X`
3. The subsequent request respect the `Set-Cookie` and send `X-TeamCity-Node-Id-Cookie=node-X` cookie

Probably, it's not so bad to leave node-id as configurable builder option too. In this case, the client can ignore `Set-Cookie` always sending explicitly specified node-id

Reply:
I was under impression that bind to a specific node is the main use case for you folks. Supporting the server-initiated pinning is a good idea, will do that. Also probably renaming binding -> pinning as it seems a bit clearer

Reply:
Server-side pinning/binding is what we need in our case - our user added to 'heavy api users' group, and we (clients) know nothing about what's happening inside, but internally TeamCity server redirects us to `node-4` using cookies. That's our use case